File: Programming_cheat_sheets/_scripting_languages/Javascript/_algorithms/quick_sort/_quickSort.js
/**
 * Quick-sort algorithm
 * https://www.youtube.com/watch?v=aQiWF4E8flQ
 * https://www.youtube.com/watch?v=COk73cpQbFQ
 *
 * quickSort can sort a specific section of an array, or the whole array.
 *
 *
 * How the algorithms works:  
 *      - most of the work is done in the partitioning function, 
 *        the 'wall' variable starts at the first index of the section of the array
 *        and you also have a 'curentIndex' variable, that will loop from the 
 *        first index of the section of the array until the last index of the section of the array,
 *        and if array[curentIndex]'s value is less than the pivot, array[wall] is 
 *        swapped with array[curentIndex] and wallIndex is incremented,
 *        so at the end, all elements less than the pivot are in the left, 
 *        and the elements greater than the pivot are in the left.
 *
 * array -> the array to be sorted
 * start -> the starting index of the array
 * end -> the end index of the array
 */
function quickSort(array, start, end){
    var pivotIndex;
    if(start < end){
        pivotIndex = partition(array, start, end);

        /**
         * recursively sort the left side and right side of the array, excluding the pivot,
         * since it is already in it's proper place.
         */
        quickSort(array, start, pivotIndex - 1);
        quickSort(array, pivotIndex + 1, end);
    }
}

/**
 * partition is an auxiliary function to the quickSort.
 * 
 * array -> the array to be partitioned
 * start -> the index to start
 * end -> the index to end
 */
function partition(array, start, end){
    var temp,
        i, 
        pivot = array[end], // The value of the pivot, used for comparison.
        wallIndex = start, // index
        currentIndex = start; // index
    
    for(i = start; i < end; i++){
        if(array[i] <= pivot){

            // swap
            temp = array[wallIndex];
            array[wallIndex] = array[i];
            array[i] = temp;

            wallIndex++; // Whatever is left from the wall is less than the pivot.
        }
    }

    // swap
    temp = array[wallIndex];
    array[wallIndex] = array[end];
    array[end] = temp;

    return wallIndex; // it's actually the pivotIndex now, because of the swap above
}

array = [5,4,6,3,7,2,8,1,9,0]

quickSort(array, 0, 9);

console.log(array); // [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ]

File: Programming_cheat_sheets/_scripting_languages/Ruby/data_structures/sorting algorithms/quick_sort.rb
# Quick sort
# https://repl.it/languages/ruby
#
# Important, this way is the easier way, using the `wall_index`
#


#
# Since quick_sort is a recursive function, we need to pass the 
# `from_index` and `to_index`.
#
def quick_sort(array, from_index, to_index)

    # This is important, we dont want to have a `from_index` greater than the `to_index`
    if from_index < to_index 

        # Get the pivot_index (which is sorted), 
        # so we can sort the left and right sides of the array
        pivot_index = partition(array, from_index, to_index)

        # in these indexes, we will not include the pivot_index, since its sorted.
        quick_sort(array, from_index, pivot_index - 1)
        quick_sort(array, pivot_index + 1, to_index)
    end
end

#
# Helper function
# it returns the index of the pivot,
# at this point, that index is in the correct position
#
def partition(array, from_index, to_index)
    pivot_value = array[to_index]
    
    wall_index = from_index

    
    # From first to second to last
    from_index.upto(to_index - 1) do |current_index|
        if array[current_index] <= pivot_value
            
            # swap
            temp = array[wall_index]
            array[wall_index] = array[current_index]
            array[current_index] = temp
            
            # At this point, the first element, upto and including `wall_index`
            # are less than the `pivot_value`, and all elements 
            # after `wall_index` upto and including `current_index` are greater than the pivot

            # lets increment the `wall_index`
            wall_index += 1
        end
    end

    # At this point all elements before and NOT including `wall_index` are 
    # leseer than the `pivot_value`
    # And all elements after and including `wall_index` are greater than the `pivot_value`.

    # So swap  `wall_index` with the `to_index`.
    temp = array[wall_index]
    array[wall_index] = array[to_index]
    array[to_index] = temp

    # Now the pivot is in its correct position!

    # Return the index of the current pivot.
    return wall_index
end


array = [5,4,10,6,3,7,0,12,2,8,1,11,9]
quick_sort(array, 0, array.size - 1)
puts array
